### PR TITLE
Navigation supports multiple jump drives

### PR DESCRIPTION
**Pull request by alextd**
**Feature**

## Summary
DistanceMap now uses dynamic fuel costs for each jump instead of a flat cost, if you have multiple jump drives.

Now, this is not going to affect vanilla as there is only the one jump drive, but most of the game code handles a ship having multiple jump drives. If one drive has a longer range, it's used when it's needed. If another drive has lower fuel use, it's used when it can be. DistanceMap didn't consider this when plotting a course, which created suboptimal paths.

DistanceMap used to use the longest range but cheapest fuel cost. So it would underestimate the cost of longer jumps, and therefore always use them. The screenshots show a clear case how this PR improves the path to a cheaper one.

Now DistanceMap will check whichever drive you have works best for each link. This includes hyperdrives - so the code doesn't need to double up on the call to Propagate(). That did require some shuffling of edgecase code in CheckLink().

## Screenshots
Before: Two long jumps => 2 * big cost.
image
After: One long jump, one short jump => 1 big + 1 small cost
image

## Usage examples
This allows multiple jump drives with varying costs/ranges to work better with plotting courses.

## Testing Done
Flew around. Checked that the flight paths made sense. Made a change that hung the game, fixed that.

## Testing NOT done
DistanceMap is also used for mission creation, Local Map, etc. which should not be affected since jump range will be 0. There are only a few missions that use the jump drive pathing, but that still should check if a hyperdrive link is available and use the lower, default hyperdrive fuel when it can.

When you haven't discovered the full map, there's edge cases that had to be handled differently, based on whether or not you know you can go from unexplored system to another. It's that chunk in CheckLink. The logic checks out, I think.

## Save File
This save file can be used to test these changes:
(Edited the player ship to have attributes of a jump drive with range 500, cost 500)
[Io Blackest of all cats.txt](https://github.com/user-attachments/files/17535661/Io.Blackest.of.all.cats.txt)

## Wiki Update
The wiki actually notes this already.

## Performance Impact
Slightly slower checking CheckLink before HasBetter, but negligible.

### CODE REVIEW

**Review by tibetiroka — COMMENTED** [review]
Seems to work really well.

**Comment by tibetiroka** [comment]
Ah, the default fuel costs were removed, and the cost is now always set in the `hyperdrive fuel` and `jump drive fuel` attributes.

**Comment by TomGoodIdea** [comment]
If you need it here too, consider moving it to a separate header, so that if someone wants to change the defaults, they would be in just one place.

File: source/Outfit.cpp [hunk 1]
```diff
@@ -28,6 +28,10 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 
 using namespace std;
 
+constexpr double Outfit::DEFAULT_HYPERDRIVE_COST = 100.;
+constexpr double Outfit::DEFAULT_SCRAM_DRIVE_COST = 150.;
+constexpr double Outfit::DEFAULT_JUMP_DRIVE_COST = 200.;
+
```
Comment (TomGoodIdea) [inline source/Outfit.cpp:34]: just initialize them in the header
```suggestion
```

File: source/Outfit.h [hunk 2]
```diff
@@ -43,6 +43,10 @@ class Outfit : public Weapon {
 	// These are all the possible category strings for outfits.
 	static const std::vector<std::string> CATEGORIES;
 
+	static const double DEFAULT_HYPERDRIVE_COST;
+	static const double DEFAULT_SCRAM_DRIVE_COST;
+	static const double DEFAULT_JUMP_DRIVE_COST;
```
Comment (TomGoodIdea) [inline source/Outfit.h:48]:
```suggestion
	static constexpr double DEFAULT_HYPERDRIVE_COST = 100.;
	static constexpr double DEFAULT_SCRAM_DRIVE_COST = 150.;
	static constexpr double DEFAULT_JUMP_DRIVE_COST = 200.;
```

File: source/Outfit.h [hunk 3]
```diff
@@ -43,6 +43,10 @@ class Outfit : public Weapon {
...
+
```
Comment (TomGoodIdea) [inline source/Outfit.h:49]:
```suggestion


```

**Comment by alextd** [comment]
I'd ask if this branch changes the default behavior of your game? It should only affect if you've got two jump drives, and in this case, one drive with very low range (i.e. cheaper use of hyperspace lanes). 

But, yea, before this branch, the 4-day path with 3 jumps could have been underestimated on fuel costs and therefore preferred with less days. Now that the longer path is properly estimated to have lower fuel cost, it is preferred for lower fuel cost. So it's technically a bugfix that exposes this design choice that fuel is more important than days.

Should it change to prefer less days instead of less fuel? Well, I'd say that depends what stage of the game you're in. Early on, I'm more worried about fuel than days... Of course, early on, there aren't going to be options to even consider. So. Sure. Yea.

**Review by tibetiroka — APPROVED** [review]

**Pull request merged**